tempesta_fw/client.c

```diff
@@ -28,121 +28,117 @@
 #include "connection.h"
 #include "log.h"
 #include "procfs.h"
+#include "tdb.h"
+#include "lib/str.h"
+#include "lib/common.h"
 
-#define CLI_HASH_BITS	17
-#define CLI_HASH_SZ	(1 << CLI_HASH_BITS)
+static struct {
+	unsigned int db_size;
+	const char *db_path;
+	unsigned int lifetime;
+} client_cfg __read_mostly;
 
 typedef struct {
-	struct hlist_head	list;
-	spinlock_t		lock;
-} CliHashBucket;
+	TfwClient	cli;
+	// TODO #515 employ eviction strategy for the table
+	time_t		expires;
+} TfwClientEntry;
 
-/*
- * TODO probably not the best container for the task and
- * HTrie should be used instead.
- */
-CliHashBucket cli_hash[CLI_HASH_SZ] = {
-	[0 ... (CLI_HASH_SZ - 1)] = {
-		HLIST_HEAD_INIT,
-	}
-};
-
-static struct kmem_cache *cli_cache;
-
-/* Total number of created clients. */
-static atomic64_t act_cli_n = ATOMIC64_INIT(0);
+static TDB *client_db;
 
 /**
  * Called when a client socket is closed.
  */
 void
 tfw_client_put(TfwClient *cli)
 {
+	TfwClientEntry *ent;
+
 	TFW_DBG2("put client %p, conn_users=%d\n",
 		 cli, atomic_read(&cli->conn_users));
 
 	if (!atomic_dec_and_test(&cli->conn_users))
 		return;
 
-	spin_lock(cli->hb_lock);
-
-	if (atomic_read(&cli->conn_users)) {
-		spin_unlock(cli->hb_lock);
-		return;
-	}
-	BUG_ON(!list_empty(&cli->conn_list));
-
-	hlist_del(&cli->hentry);
-
-	spin_unlock(cli->hb_lock);
+	ent = (TfwClientEntry *)cli;
+	ent->expires = tfw_current_timestamp() + client_cfg.lifetime;
 
 	TFW_DBG("free client: cli=%p\n", cli);
-	kmem_cache_free(cli_cache, cli);
-	atomic64_dec(&act_cli_n);
 	TFW_DEC_STAT_BH(clnt.online);
 }
 EXPORT_SYMBOL(tfw_client_put);
 
+static bool
+addr_eq(TdbRec *rec, void *data)
+{
+	TfwClientEntry *ent = (TfwClientEntry *)rec->data;
+	TfwClient *cli = &ent->cli;
+
+	return !memcmp_fast(&cli->addr, data, sizeof(cli->addr));
```

Comment:
The function is used to find the `TfwClient` structure matching to the client's address. But all fields of `TfwAddr` are matched, including port. This will allocate a new `TfwClient` structure for connection with the same ip, but different source port. It's incorrect and insecure behaviour. Bots will just open a new connection, but we will treat it as a new client.

Previously, only IP address was used to match `TfwClient` structure for the client.